sklearn/tree/_splitter.pyx

```diff
@@ -220,222 +221,178 @@ cdef class Splitter:
-                    continue
-
-                current.pos = p
-
-                # Reject if min_samples_leaf is not guaranteed
-                if (((current.pos - start) < min_samples_leaf) or
-                        ((end - current.pos) < min_samples_leaf)):
-                    continue
+            criterion.update(current.pos)
 
-                self.criterion.update(current.pos)
-
-                # Reject if min_weight_leaf is not satisfied
-                if ((self.criterion.weighted_n_left < min_weight_leaf) or
-                        (self.criterion.weighted_n_right < min_weight_leaf)):
-                    continue
-
-                current_proxy_improvement = self.criterion.proxy_impurity_improvement()
-
-                if current_proxy_improvement > best_proxy_improvement:
-                    best_proxy_improvement = current_proxy_improvement
-                    # sum of halves is used to avoid infinite value
-                    current.threshold = Xf[p - 1] / 2.0 + Xf[p] / 2.0
-
-                    if (
-                        current.threshold == Xf[p] or
-                        current.threshold == INFINITY or
-                        current.threshold == -INFINITY
-                    ):
-                        current.threshold = Xf[p - 1]
-
-                    best = current  # copy
-
-        # Reorganize into samples[start:best.pos] + samples[best.pos:end]
-        if best.pos < end:
-            partition_end = end
-            p = start
-
-            while p < partition_end:
-                if self.X[samples[p], best.feature] <= best.threshold:
-                    p += 1
-
-                else:
-                    partition_end -= 1
-
-                    samples[p], samples[partition_end] = samples[partition_end], samples[p]
+            # Reject if min_weight_leaf is not satisfied
+            if ((criterion.weighted_n_left < min_weight_leaf) or
+                    (criterion.weighted_n_right < min_weight_leaf)):
+                continue
 
-            self.criterion.reset()
-            self.criterion.update(best.pos)
-            self.criterion.children_impurity(&best.impurity_left,
-                                             &best.impurity_right)
-            best.improvement = self.criterion.impurity_improvement(
-                impurity, best.impurity_left, best.impurity_right)
+            current_proxy_improvement = criterion.proxy_impurity_improvement()
 
-        # Respect invariant for constant features: the original order of
-        # element in features[:n_known_constants] must be preserved for sibling
-        # and child nodes
-        memcpy(&features[0], &constant_features[0], sizeof(SIZE_t) * n_known_constants)
+            if current_proxy_improvement > best_proxy_improvement:
+                best_proxy_improvement = current_proxy_improvement
+                # sum of halves is used to avoid infinite value
+                current.threshold = Xf[p_prev] / 2.0 + Xf[p] / 2.0
```

Comment:
We would not avoid a division in this case? We might introduce a different rounding error thought.
```suggestions
                current.threshold = (Xf[p_prev] + Xf[p]) / 2.0
```

Reply:
We can skip this because this is more than a refactoring.